### ISSUE TITLE
With reticulate, xgboost with certain type of hyperparameters cannot be paralleled.

### ISSUE THREAD

**Issue by flippercy** [rstudio/reticulate#996]
I am building a random forest via xgboost following the instruction at https://xgboost.readthedocs.io/en/latest/tutorials/rf.html. The R codes are like below (using the sample from the library xgboost) :

data(agaricus.train, package='xgboost')
data(agaricus.test, package='xgboost')
train <- agaricus.train
test <- agaricus.test
random.forest.model<- xgboost(data = train$data, label = train$label, max_depth = 6,
eta = 1, nthread = 5, nrounds = 1, objective = "binary:logistic", num_parallel_tree = 10000)

When running the codes above in RStudio directly, from the backend of the server (centos Linux) I can see that five CPUs are activated and used for the model.

However, when I run the python equivalent in RStudio via reticulate, no matter how I set nthread, only one CPU is active and used, indicating that the process becomes single-threaded.

 param = {'max_depth': 6, 
                 'eta': 1,
                  'nthread': 5,
                   'num_parallel_tree': 10000,
                   'objective': 'binary:logistic'}

random_forest_model = xgb.train(param, dtrain, num_round=1)

Moreover, the python codes above can be run in parallel using Jupyter with no issue.  In addition, for a 'traditional' xgboost with multiple rounds, there is no such issue in reticulate.

Any ideas? The process takes a lot of time after becoming single-threaded, which is very unfavorable for my work.

Thank you.

**Comment by kevinushey** [comment]
Can you please share a self-contained and minimally reproducible example, so I can more easily run, reproduce and see the issue locally?

**Comment by flippercy** [comment]
@kevinushey here is an example in python with FLAML (https://github.com/microsoft/FLAML)

!pip install flaml[notebook];
from flaml.data import load_openml_dataset
from xgboost import XGBClassifier
from flaml.model import BaseEstimator
X_train, X_test, y_train, y_test = load_openml_dataset(dataset_id=1169, data_dir='./')
from flaml import AutoML
from flaml import tune
automl = AutoML()
num_cores = 16
randomseed = 19850606

class MyMonotonicRandomForestClassifier(BaseEstimator):


    def __init__(self, task = 'binary:logistic', n_jobs = num_cores, **params):

        super().__init__(task, **params)

        self.estimator_class = XGBClassifier

        # convert to int for integer hyperparameters
        self.params = {
            'n_jobs': params['n_jobs'] if 'n_jobs' in params else num_cores,
            'booster':params['booster'] if 'booster' in params else 'gbtree',
            'learning_rate': params['learning_rate'] if 'learning_rate' in params else 1,
            'gamma': params['gamma'],
            'reg_lambda': params['reg_lambda'],
            'reg_alpha': params['reg_alpha'],
            'max_depth': int(params['max_depth']),
            'min_child_weight': int(params['min_child_weight']),
            'subsample': params['subsample'],
            'colsample_bylevel':params['colsample_bylevel'],
            'num_parallel_tree':int(params['num_parallel_tree']),
            'n_estimators':params['n_estimators'] if 'n_estimators' in params else 1,
            'random_state': params['random_state'] if 'random_state' in params else randomseed
            
       }   

    @classmethod
    def search_space(cls, data_size, task):
        
        space = {        
        'max_depth': {'domain': tune.uniform(lower=15, upper=100), 'init_value': 30},
        'num_parallel_tree': {'domain': tune.uniform(lower = 5000, upper = 20000), 'init_value': 10000},
        'min_child_weight': {'domain': tune.uniform(lower = 1, upper = 1000), 'init_value': 100},
        'subsample': {'domain': tune.uniform(lower = 0.5, upper = 1), 'init_value': 0.67},
        'colsample_bylevel': {'domain': tune.uniform(lower = 0.7, upper = 1), 'init_value': 0.9},
        'gamma': {'domain': tune.loguniform(lower = 0.000000000001, upper = 0.001), 'init_value': 0.00001},
        'reg_lambda': {'domain': tune.loguniform(lower = 0.000000000001, upper = 1), 'init_value': 1},
         'reg_alpha': {'domain': tune.loguniform(lower = 0.000000000001, upper = 1), 'init_value': 0.000000000001},
        }
        return space
    
automl.add_learner(learner_name = 'MonotonicRandomForest', learner_class = MyMonotonicRandomForestClassifier)


settings = {
    "time_budget": 300,  # total running time in seconds
    "metric": 'roc_auc', 
    "estimator_list": ['MonotonicRandomForest'],  # list of ML learners
    "task": 'classification',  # task type    
    "log_file_name": 'airlines_experiment_custom.log',  # flaml log file 
    "log_training_metric": True,  # whether to log training metric
}

automl.fit(X_train = X_train, y_train = y_train, 
X_val = X_test, y_val = y_test, **settings)


When running the codes above in Jupyter, 16 CPUs in my linux server are activated and used; however, when I put the codes in RStudio and run via reticulate in the same server, only one CPU was activated as used (Seems that 16 CPUs were activated at the beginning but quickly all went idle but one.

Let me know if you need anything else.

Thank you.
